File: tests/controller/mf.py
#!/usr/bin/env python
'''
 @file mf.py
 @brief Testing membership function
 @copyright Copyright (C) 2020 tqfx, All rights reserved.
'''
import os
import sys

sys.path.insert(0, os.getcwd())
prefix = os.path.join(sys.path[0], "build")
if not os.path.exists(prefix):
    os.mkdir(prefix)
try:
    import a
    import numpy as np
    import matplotlib.pyplot as plt
except Exception as e:
    print(e)
    exit()

title = "Gaussian membership function"
x = np.arange(-3, 3, 0.001)
y = a.mf.gauss(x, 1, 0)
plt.figure(title)
plt.title(title)
plt.plot(x, y)
plt.grid(True)
plt.savefig(os.path.join(prefix, "mf_gauss.png"))

title = "Generalized bell-shaped membership function"
x = np.arange(-3, 3, 0.001)
y = a.mf.gbell(x, 2, 1, 0)
plt.figure(title)
plt.title(title)
plt.plot(x, y)
plt.grid(True)
plt.savefig(os.path.join(prefix, "mf_gbell.png"))

title = "Sigmoidal membership function"
x = np.arange(-3, 3, 0.001)
y = a.mf.sig(x, 2, 0)
plt.figure(title)
plt.title(title)
plt.plot(x, y)
plt.grid(True)
plt.savefig(os.path.join(prefix, "mf_sig.png"))

title = "Triangular membership function"
x = np.arange(0, 2, 0.001)
y = a.mf.tri(x, 0, 1, 2)
plt.figure(title)
plt.title(title)
plt.plot(x, y)
plt.grid(True)
plt.savefig(os.path.join(prefix, "mf_tri.png"))

y = []
title = "Trapezoidal membership function"
x = np.arange(0, 3, 0.001)
y = a.mf.trap(x, 0, 1, 2, 3)
plt.figure(title)
plt.title(title)
plt.plot(x, y)
plt.grid(True)
plt.savefig(os.path.join(prefix, "mf_trap.png"))

title = "Z-shaped membership function"
x = np.arange(0, 3, 0.001)
y = a.mf.z(x, 1, 2)
plt.figure(title)
plt.title(title)
plt.plot(x, y)
plt.grid(True)
plt.savefig(os.path.join(prefix, "mf_z.png"))

params = [
    (a.mf.TRI, -3, -3, -2),
    (a.mf.TRI, -3, -2, -1),
    (a.mf.TRI, -2, -1, +0),
    (a.mf.TRI, -1, +0, +1),
    (a.mf.TRI, +0, +1, +2),
    (a.mf.TRI, +1, +2, +3),
    (a.mf.TRI, +2, +3, +3),
]

title = "membership function"
x = np.arange(-3, 3, 0.001)
plt.figure(title)
plt.title(title)
for param in params:
    plt.plot(x, a.mf()(param[0], x, param[1:]))
plt.savefig(os.path.join(prefix, "mf.png"))

File: tests/controller/mf.c
/*!
 @file mf.c
 @brief test membership function
 @copyright Copyright (C) 2020 tqfx, All rights reserved.
*/

#include "a/controller/mf.h"

static void test(void)
{
    a_mf_gauss(0, 1, 0);
    a_mf_gbell(0, 1, 1, 0);
    a_mf_sig(0, 1, 0);
    a_mf_trap(0, 0, 1, 2, 3);
    a_mf_tri(0, 0, 1, 2);
    a_mf_z(0, 0, 1);
    a_real_t p[] = {1, 2, 3, 4};
    a_mf(A_MF_NONE, 2, p);
    a_mf(A_MF_GAUSS, 2, p);
    a_mf(A_MF_GBELL, 2, p);
    a_mf(A_MF_SIG, 2, p);
    a_mf(A_MF_TRAP, 2, p);
    a_mf(A_MF_TRI, 2, p);
    a_mf(A_MF_Z, 2, p);
}

int main(void)
{
    test();
    return 0;
}
